iommi/sql_trace.py

```diff
@@ -0,0 +1,401 @@
+                    _first = False
+                    fg = BASE_COLORS[token]
+                    using = getattr(record, 'using', None)
+                    if using == 'read-only':
+                        fg = 'white'
+                    if using == 'vertica':
+                        fg = 'blue'
+                    if not short and fat_arrow:
+                        yield format_html('<span>{}<br>&nbsp;</span>', colorize('===>', fg=fg))
+                        column = 2
+                if not short:
+                    if token in ('AND', 'OR', 'LEFT', 'INNER', 'FROM', 'WHERE', 'ORDER', 'LIMIT'):
+                        yield format_html('<span><br>&nbsp;</span>')
+                        column = 2
+                        if token in ('AND', 'OR',):
+                            yield format_html('&nbsp;')
+                            column += 2
+            yield colorize(token, fg=fg, bold=bold)
+
+            column += len(token) + 1
+            if column > width and token[-1] == ',':
+                yield format_html('<span><br>{}</span>', format_html('&nbsp;' * 8))
+                column = 8
+        if duration:
+            yield format_html(f"<span>-- [{duration:.3f}s]</span>")
+
+    tokens = list(tokenize(text))
+    return format_html('<span>' + ('{}' * len(tokens)) + '</span>', *tokens)
+
+
+def safe_unicode_literal(params):
+    if params is None:
+        return 'NULL'
+
+    if isinstance(params, (list, tuple)):
+        return tuple([safe_unicode_literal(x) for x in params])
+    elif isinstance(params, (float, int)):
+        return repr(params)
+    elif isinstance(params, (date, datetime)):
+        return repr(params.isoformat())
+    elif isinstance(params, dict):
+        return dict((k, safe_unicode_literal(v)) for k, v in params.items())
+    elif isinstance(params, bytes):
+        return repr(params.decode(errors="replace"))
+    else:
+        return repr(params)
+
+
+state = threading.local()
+
+
+def set_sql_debug(new_state, validate=True):
+    if new_state == 'None':
+        new_state = None
+    if validate:
+        assert new_state is None or new_state in SQL_DEBUG_LEVELS
+    setattr(state, 'sql_debug', new_state)
+
+
+def get_sql_debug():
+    """ :rtype: bool """
+    result = getattr(state, 'sql_debug', None)
+    if result is not None:
+        return result
+    return getattr(settings, 'SQL_DEBUG', False)
+
+
+@contextmanager
+def no_sql_debug():
+    """
+    Context manager to temporarily suspend sql logging.
+
+    This is useful inside the sql debug implementation to avoid infinite recursion.
+    """
+    old_state = get_sql_debug()
+    set_sql_debug(False)
+    yield
+    set_sql_debug(old_state)
+
+
+SQL_DEBUG_LEVEL_ALL = 'all'
+SQL_DEBUG_LEVEL_ALL_WITH_STACKS = 'stacks'
+SQL_DEBUG_LEVEL_WORST = 'worst'
+
+
+SQL_DEBUG_LEVELS = {
+    SQL_DEBUG_LEVEL_ALL,
+    SQL_DEBUG_LEVEL_ALL_WITH_STACKS,
+    SQL_DEBUG_LEVEL_WORST,
+    False,
```

Comment:
A bit weird to have a bool among the strings. Would not a setting to None be more natural? (So the field is optional.)

Reply:
Fixed